File: src/types/enums.ts

import { t as uiText } from "./../lib/i18n/index.ts";
// ============================================================================
// PerfectPPI Enums — mirrors database enums + client-only types
// ============================================================================

// Domain A: Identity
export type UserRole =
  | "consumer"
  | "technician"
  | "org_manager"
  | "admin"
  | "developer";

/**
 * Roles a developer account can switch itself into, in the order they are
 * offered in settings. "developer" is included so there is a way back to the
 * switcher's own portal after landing in another role.
 */
export const SWITCHABLE_ROLES = [
  "developer",
  "consumer",
  "technician",
  "org_manager",
  "admin",
] as const satisfies readonly UserRole[];

export const USER_ROLE_LABELS: Record<UserRole, string> = {
  consumer: uiText("ui.consumer_3fdb185870"),
  technician: uiText("ui.technician_9041ccc417"),
  org_manager: uiText("ui.organization_manager_e5da2451a1"),
  admin: uiText("ui.admin_c1c224b03c"),
  developer: uiText("ui.developer_3fb7b39416"),
};

export const USER_ROLE_DESCRIPTIONS: Record<UserRole, string> = {
  consumer: uiText("ui.vehicles_listings_inspection_requests_warran_37b9348016"),
  technician: uiText("ui.assigned_inspection_queue_submissions_review_120633d424"),
  org_manager: uiText("ui.organization_roster_inspections_dealerspace_17ee364020"),
  admin: uiText("ui.platform_wide_moderation_outputs_contracts_a_fc93692c0a"),
  developer: uiText("ui.this_switcher_no_portal_data_of_its_own_a5cd5d08fd"),
};

// Domain B: Vehicles
export type VehicleVisibility = "public" | "friends" | "private";
export type MediaType = "image" | "video";
export type CommunityPostAudience = "public" | "friends";

// Domain C: PPI Engine
export type WhoseCar = "own" | "other";
export type RequesterRole = "buying" | "selling" | "documenting";
export type PerformerType = "self" | "technician";

export type PpiType = "personal" | "general_tech" | "certified_tech";

/**
 * Which question set an inspection asks. Distinct from `PpiType`, which is the
 * trust tier derived from the performer's certification.
 */
export type InspectionScope = "complete" | "dents_tires";

export type PpiRequestStatus =
  | "draft"
  | "pending_assignment"
  | "assigned"
  | "accepted"
  | "in_progress"
  | "submitted"
  | "needs_revision"
  | "completed"
  | "archived";

export type PpiSubmissionStatus =
  | "draft"
  | "in_progress"
  | "submitted"
  | "completed";

export type SectionType =
  | "vehicle_basics"
  | "dashboard_warnings"
  | "exterior"
  | "interior"
  | "engine_bay"
  | "tires_brakes"
  | "suspension_steering"
  | "fluids"
  | "electrical_controls"
  | "underbody"
  | "road_test"
  | "modifications"
  // dents_tires scope only
  | "wheels_tires"
  | "body_damage";

export type CompletionState = "not_started" | "in_progress" | "completed";
export type AnswerType =
  | "text"
  | "yes_no"
  | "select"
  | "number"
  // Catalog 2 structured observations; the value lives in ppi_answers.observation.
  | "measurement"
  | "tire_markings"
  | "dot_code"
  | "condition_scale"
  | "defect_list"
  | "tire_placard"
  | "panel_condition";

// Domain D: Warranty
export type WarrantyOptionStatus =
  | "not_offered"
  | "offered"
  | "viewed"
  | "selected";

export type WarrantyOrderStatus =
  | "contract_pending"
  | "signed"
  | "payment_pending"
  | "paid"
  | "failed"
  | "cancelled";

export type PaymentMethod = "card" | "bank_transfer" | "financing";
export type PaymentStatus = "pending" | "completed" | "failed" | "refunded";

// Domain E: Communications
export type MessageStatus = "unread" | "read" | "archived";
export type ShareTargetType =
  | "media_package"
  | "inspection_result"
  | "standardized_output";

// Domain F: Technician
export type CertificationLevel = "none" | "ase" | "master" | "oem_qualified";
export type OrgMemberRole = "technician" | "manager";

// Domain G: Notifications
export type NotificationType =
  | "friend_request"
  | "friend_request_accepted"
  | "comment_reply"
  | "post_comment"
  | "post_likes"
  | "post_mention"
  | "answer_accepted"
  | "accepted_answer_unavailable"
  | "answer_helpful"
  | "group_post_removed"
  | "group_role_changed"
  | "group_invitation"
  | "group_join_request"
  | "group_join_decision"
  | "saved_listing_updated"
  | "saved_search_match"
  | "build_update"
  | "event_cancelled"
  | "event_update"
  | "listing_inspection_requested"
  | "moderation_decision"
  | "moderation_case"
  | "report_received"
  | "tech_request_new"
  | "tech_request_accepted"
  | "inspection_submitted"
  | "inspection_updated"
  | "warranty_available"
  | "payment_completed"
  | "message_received";

// Domain G: Audit
export type AuditAction =
  | "inspection_edited"
  | "output_regenerated"
  | "contract_state_changed"
  | "payment_state_changed"
  | "submission_resubmitted"
  | "output_review_released";

// Factual inspection-source labels. Avoid ranking or endorsement language.
export const PPI_TRUST_TIERS = {
  personal: { label: uiText("ui.personal_inspection_70f8e0f2cb"), badge: uiText("ui.owner_entered_5b48a5e342"), color: "amber" },
  general_tech: { label: uiText("ui.technician_inspection_cd51204a54"), badge: uiText("ui.technician_9041ccc417"), color: "slate" },
  certified_tech: {
    label: uiText("ui.technician_inspection_cd51204a54"),
    badge: uiText("ui.reviewed_credential_3240dd1055"),
    color: "yellow",
  },
} as const;

File: mobile-app/PerfectPPI/Core/Models/Enums.swift
import Foundation

// Mirrors src/types/enums.ts on the server. Keep in sync.

enum UserRole: String, Codable, CaseIterable {
    case consumer
    case technician
    case orgManager = "org_manager"
    case admin
    // A developer account parks here between roles. It has no tab bar of its
    // own — the switcher stands in for one. It must decode regardless:
    // Profile.role is a RawRepresentable, so an unrecognized string fails the
    // whole profile decode and locks the account out of the app entirely.
    case developer

    /// Roles a developer can switch into, in the order the switcher lists them.
    /// Mirrors SWITCHABLE_ROLES in src/types/enums.ts.
    static let switchable: [UserRole] = [
        .developer, .consumer, .technician, .orgManager, .admin,
    ]

    var label: String {
        switch self {
        case .consumer: return "Consumer"
        case .technician: return "Technician"
        case .orgManager: return "Organization Manager"
        case .admin: return "Admin"
        case .developer: return "Developer"
        }
    }

    var summary: String {
        switch self {
        case .consumer: return "Vehicles, listings, inspection requests, warranties."
        case .technician: return "Assigned inspection queue, submissions, reviews."
        case .orgManager: return "Organization roster, inspections, DealerSpace."
        case .admin: return "Platform-wide moderation, outputs, audit log."
        case .developer: return "This switcher. No portal data of its own."
        }
    }

    var icon: String {
        switch self {
        case .consumer: return "car.fill"
        case .technician: return "wrench.and.screwdriver.fill"
        case .orgManager: return "building.2.fill"
        case .admin: return "shield.lefthalf.filled"
        case .developer: return "hammer.fill"
        }
    }
}

enum MediaType: String, Codable { case image, video }
enum VehicleVisibility: String, Codable, CaseIterable, Identifiable {
    case `public`, friends, `private`

    var id: String { rawValue }

    var label: String {
        switch self {
        case .public: "Public"
        case .friends: "Friends"
        case .private: "Only me"
        }
    }

    var systemImage: String {
        switch self {
        case .public: "globe"
        case .friends: "person.2.fill"
        case .private: "lock.fill"
        }
    }
}
enum VehicleOwnershipState: String, Codable, CaseIterable, Identifiable {
    case owned
    case previouslyOwned = "previously_owned"
    case considering
    case project

    var id: String { rawValue }

    var label: String {
        switch self {
        case .owned: "Owned"
        case .previouslyOwned: "Previously owned"
        case .considering: "Shopping / considering"
        case .project: "Project"
        }
    }
}

enum WhoseCar: String, Codable { case own, other }
enum RequesterRole: String, Codable { case buying, selling, documenting }
enum PerformerType: String, Codable { case selfInspection = "self", technician }

enum PpiType: String, Codable {
    case personal
    case generalTech = "general_tech"
    case certifiedTech = "certified_tech"
}

enum PpiRequestStatus: String, Codable {
    case draft
    case pendingAssignment = "pending_assignment"
    case assigned
    case accepted
    case inProgress = "in_progress"
    case submitted
    case needsRevision = "needs_revision"
    case completed
    case archived
}

enum PpiSubmissionStatus: String, Codable {
    case draft
    case inProgress = "in_progress"
    case submitted
    case completed
}

/// Which question set an inspection asks. Distinct from `PpiType`, which is the
/// trust tier derived from the performer's certification.
enum InspectionScope: String, Codable, CaseIterable {
    case complete
    case dentsTires = "dents_tires"

    var label: String {
        switch self {
        case .complete: return "Complete Inspection"
        case .dentsTires: return "Dents & Tires"
        }
    }

    var summary: String {
        switch self {
        case .complete: return "Full pre-purchase inspection — every system, inside and out."
        case .dentsTires: return "Tire tread, wheels, and cosmetic body damage."
        }
    }
}

enum SectionType: String, Codable, CaseIterable {
    case vehicleBasics = "vehicle_basics"
    case dashboardWarnings = "dashboard_warnings"
    case exterior
    case interior
    case engineBay = "engine_bay"
    case tiresBrakes = "tires_brakes"
    case suspensionSteering = "suspension_steering"
    case fluids
    case electricalControls = "electrical_controls"
    case underbody
    case roadTest = "road_test"
    case modifications
    // dents_tires scope only
    case wheelsTires = "wheels_tires"
    case bodyDamage = "body_damage"

    var label: String {
        switch self {
        case .wheelsTires: return "Wheels & Tires"
        default: return rawValue.replacingOccurrences(of: "_", with: " ").capitalized
        }
    }
}

enum CompletionState: String, Codable {
    case notStarted = "not_started"
    case inProgress = "in_progress"
    case completed
}

enum AnswerType: String, Codable {
    case text
    case yesNo = "yes_no"
    case select
    case number
}

enum WarrantyOrderStatus: String, Codable {
    case contractPending = "contract_pending"
    case signed
    case paymentPending = "payment_pending"
    case paid
    case failed
    case cancelled
}

enum PaymentStatus: String, Codable {
    case pending, completed, failed, refunded
}

enum CertificationLevel: String, Codable {
    case none, ase, master
    case oemQualified = "oem_qualified"
}

enum NotificationType: String, Codable {
    case techRequestNew = "tech_request_new"
    case techRequestAccepted = "tech_request_accepted"
    case inspectionSubmitted = "inspection_submitted"
    case inspectionUpdated = "inspection_updated"
    case warrantyAvailable = "warranty_available"
    case paymentCompleted = "payment_completed"
    case messageReceived = "message_received"
    case moderationDecision = "moderation_decision"
    case moderationCase = "moderation_case"
    case reportReceived = "report_received"
    case answerAccepted = "answer_accepted"
    case acceptedAnswerUnavailable = "accepted_answer_unavailable"
    case listingInspectionRequested = "listing_inspection_requested"
    case friendRequest = "friend_request"
    case friendRequestAccepted = "friend_request_accepted"
    case postComment = "post_comment"
    case postLikes = "post_likes"
    case postMention = "post_mention"
    case groupPostRemoved = "group_post_removed"
    case groupRoleChanged = "group_role_changed"
    case groupInvitation = "group_invitation"
    case groupJoinRequest = "group_join_request"
    case groupJoinDecision = "group_join_decision"
    case savedListingUpdated = "saved_listing_updated"
    /// Server-side categories grow over time (plan 22.1); an unrecognized
    /// value must not fail decoding of the whole inbox on an older build.
    case unknown

    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = NotificationType(rawValue: raw) ?? .unknown
    }
}

enum ListingStatus: String, Codable {
    case active, sold, archived
}

enum CommunityContentStatus: String, Codable {
    case active, hidden, archived
}

enum CommunityPostType: String, Codable, CaseIterable {
    case general
    case question
    case buildUpdate = "build_update"
    case maintenance
    case beforeAfter = "before_after"
    case inspectionDiscussion = "inspection_discussion"
    case buyingAdvice = "buying_advice"
    case poll
    /// Newer server types never fail feed decoding on an older build.
    case unknown

    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = CommunityPostType(rawValue: raw) ?? .unknown
    }
}

enum CommunityFeedFilter: String, CaseIterable, Identifiable {
    case all
    case friends
    case myCars = "my_cars"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .all: "All"
        case .friends: "Friends"
        case .myCars: "My Cars"
        }
    }
}

enum MentionPolicy: String, Codable, CaseIterable, Identifiable {
    case everyone
    case friendsAndGroups = "friends_and_groups"
    case friends
    case nobody

    var id: String { rawValue }

    var label: String {
        switch self {
        case .everyone: "Everyone who can see the post"
        case .friendsAndGroups: "Friends and group members"
        case .friends: "Friends only"
        case .nobody: "Nobody"
        }
    }
}

enum CommunityPostAudience: String, Codable, CaseIterable, Identifiable {
    case `public`, friends

    var id: String { rawValue }
    var label: String { self == .public ? "Public" : "Friends" }
}

enum ReviewStatus: String, Codable {
    case active, hidden
}

enum ShareTargetType: String, Codable {
    case mediaPackage = "media_package"
    case inspectionResult = "inspection_result"
    case standardizedOutput = "standardized_output"
}
